kornia/contrib/dsnt.py

```diff
@@ -0,0 +1,220 @@
+        temperature (torch.Tensor):
+
+    Shape:
+        - Input: :math:`(B, N, H, W)`
+        - Output: :math:`(B, N, H, W)`
+    """
+    _validate_batched_image_tensor_input(input)
+
+    batch_size, channels, height, width = input.shape
+    x: torch.Tensor = input.view(batch_size, channels, -1)
+
+    x_soft: torch.Tensor = F.softmax(x * temperature, dim=-1)
+
+    return x_soft.view(batch_size, channels, height, width)
+
+
+def spatial_softargmax_2d(
+        input: torch.Tensor,
+        normalized_coordinates: bool = True,
+) -> torch.Tensor:
+    r"""Computes the 2D soft-argmax of a given input heatmap.
+
+    The input heatmap is assumed to represent a valid spatial probability
+    distribution, which can be achieved using
+    :class:`~kornia.contrib.dsnt.spatial_softmax_2d`.
+
+    Arguments:
+        input (torch.Tensor):
+        normalized_coordinates (bool): whether to return the
+          coordinates normalized in the range of [-1, 1]. Otherwise,
+          it will return the coordinates in the range of the input shape.
+          Default is True.
+
+    Shape:
+        - Input: :math:`(B, N, H, W)`
+        - Output: :math:`(B, N, 2)`
+    """
+    _validate_batched_image_tensor_input(input)
+
+    batch_size, channels, height, width = input.shape
+
+    # Create coordinates grid.
+    grid: torch.Tensor = create_meshgrid(
+        height, width, normalized_coordinates)
+    grid = grid.to(device=input.device, dtype=input.dtype)
+
+    pos_x: torch.Tensor = grid[..., 0].reshape(-1)
+    pos_y: torch.Tensor = grid[..., 1].reshape(-1)
+
+    input_flat: torch.Tensor = input.view(batch_size, channels, -1)
+
+    # Compute the expectation of the coordinates.
+    expected_y: torch.Tensor = torch.sum(pos_y * input_flat, -1, keepdim=True)
+    expected_x: torch.Tensor = torch.sum(pos_x * input_flat, -1, keepdim=True)
+
+    output: torch.Tensor = torch.cat([expected_x, expected_y], -1)
+
+    return output.view(batch_size, channels, 2)  # BxNx2
+
+
+def render_gaussian_2d(
+        mean: torch.Tensor,
+        std: torch.Tensor,
+        size: List[int],
+        normalized_coordinates: bool = True,
+):
+    r"""Renders the PDF of a 2D Gaussian distribution.
+
+    Arguments:
+        mean (torch.Tensor): the mean location of the Gaussian to render,
+          :math:`(\mu_x, \mu_y)`.
+        std (torch.Tensor): the standard deviation of the Gaussian to render,
+          :math:`(\sigma_x, \sigma_y)`.
+        size (list): the (height, width) of the output image.
+        normalized_coordinates: whether `mean` and `std` are assumed to use
+          coordinates normalized in the range of [-1, 1]. Otherwise,
+          coordinates are assumed to be in the range of the output shape.
+          Default is True.
+
+    Shape:
+        - `mean`: :math:`(*, 2)`
+        - `std`: :math:`(*, 2)`
+        - Output: :math:`(*, H, W)`
+    """
+    device = mean.device
+    dtype = mean.dtype
+    if not dtype.is_floating_point:
```

Comment:
for mypy if I don't have any clue about this, I don't really know why is not recognised at all. If that blocks you, use ` # type: ignore` to avoid mypy check in this line

Reply:
Using `# type: ignore` doesn't work in that case, as mentioned in this note from the mypy docs:

> The # type: ignore comment will only assign the implicit Any type if mypy cannot find information about that particular module. So, if we did have a stub available for frobnicate then mypy would ignore the # type: ignore comment and typecheck the stub as usual.

However, I can annotate the type of `dtype` as Any and that does work.